### PR TITLE
lsp: fix jump list in jump_to_location

### PR DESCRIPTION
**Pull request by fsouza**
Fixes regression introduced in #12262.

### CODE REVIEW

**Comment by SanchayanMaity** [comment]
FWIW, can confirm this fixes the issue for me.

**Comment by fsouza** [comment]
@mjlbach I used this for a couple of days and it was fine, but then started running into some other issues when jumping to different files and ended-up forking that function in my config, because I didn't have much time to debug it :/ I may have time to look into it over the weekend, can't promise it though 😞 

Either way, this PR definitely fixes jumping within a buffer, I just wanted to add some tests before marking it as ready.

**Comment by gbrlsnchs** [comment]
@fsouza What kind of errors did you find with this fix? This definitely looks like it fixes the regression.

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -957,10 +957,15 @@ function M.jump_to_location(location)
   local motionless =
     bufnr == from_bufnr and
     row+1 == from[2] and col+1 == from[3]
-  if not motionless and _should_add_to_tagstack(item) then
-    local winid = vim.fn.win_getid()
-    local items = {item}
-    vim.fn.settagstack(winid, {items=items}, 't')
+  if not motionless then
+    -- Save position in jumplist
+    vim.cmd "normal! m'"
```
Comment (mhinz) [inline runtime/lua/vim/lsp/util.lua:962]: Maybe this should be `vim.cmd "mark '"` instead. No need to switch modes.

**Comment by mhinz** [comment]
FWIW, I also see incorrect behavior here. `:lua vim.lsp.buf.definition()` followed by `<c-o>` is only sometimes doing what it's supposed to do.

Still need to read up on when this regressed.

**Comment by gbrlsnchs** [comment]
> FWIW, I also see incorrect behavior here. `:lua vim.lsp.buf.definition()` followed by `<c-o>` is only sometimes doing what it's supposed to do.
> 
> Still need to read up on when this regressed.

Basically the regression stopped pushing the jump to the jumplist when it occurs in the same buffer. `vim.lsp.buf.definition` uses a utility function affected by the regression IIRC, so yeah, this might also fix the issue you witnessed.

**Comment by fsouza** [comment]
@gbrlsnchs I was seeing what @mhinz is seeing too, sometimes c-o would do what I expected, other times wouldn't. I didn't investigate exactly when nor why the issue would appear.

**Comment by gbrlsnchs** [comment]
> @gbrlsnchs I was seeing what @mhinz is seeing too, sometimes c-o would do what I expected, other times wouldn't. I didn't investigate exactly when nor why the issue would appear.

Maybe it would be enough to just put the piece of code that adds to the jumplist back to where it was before the regression and it can be improved later, what do you think?

**Comment by gbrlsnchs** [comment]
Also, I don't think the jumplist suffers from these "phantom jumps". There's [this commit](https://github.com/neovim/neovim/commit/35362495c965554f45634bcde3c4ce6d5eca52aa) that seems to have fixed it, which is understandable since the jumplist has a lot more usage than the tagstack (which I personally only use with LSP).

Therefore, this fix could be simpler by just restoring the old behavior, I guess. Of course, testing this might be important so no one removes that piece of code like I did, but just reverting to the old behavior is a good starting point IMO.

**Comment by mhinz** [comment]
I agree.

@fsouza Please change the commit to just do this:

```diff
diff --git i/runtime/lua/vim/lsp/util.lua w/runtime/lua/vim/lsp/util.lua
index 92ec447b5..ce8468aa8 100644
--- i/runtime/lua/vim/lsp/util.lua
+++ w/runtime/lua/vim/lsp/util.lua
@@ -944,6 +944,9 @@ function M.jump_to_location(location)
   local from = {from_bufnr, vim.fn.line('.'), vim.fn.col('.'), 0}
   local item = {tagname=vim.fn.expand('<cword>'), from=from}

+  -- Save position in jumplist
+  vim.cmd("mark '")
+
   --- Jump to new location (adjusting for UTF-16 encoding of characters)
   local bufnr = vim.uri_to_bufnr(uri)
   api.nvim_set_current_buf(bufnr)
```

**Comment by SanchayanMaity** [comment]
Seems I jumped to soon to declared it as fix. This MR does fix the case of jumping in the same buffer and back case as noted above, but, even with @mhinz change applied, while jumping back with C-o there seems to be one spurious jump before jumping back to correct location.

**Comment by mhinz** [comment]
@SanchayanMaity

My diff was meant to be applied to master, not to this PR. (Just in case you did that.) And the diff is really just restoring the line that was [left out in `#12262`](https://github.com/neovim/neovim/pull/12262/files#diff-de132e20a0c678f2ff3234a05d868b0d118561cb72317910e7f28839640ed2e3L663).

**Comment by SanchayanMaity** [comment]
@mhinz Oh I see. I thought that had to be applied in addition to this MR. My mistake. Just applying your one liner, works properly. Thank you.

**Comment by fsouza** [comment]
Updated the PR, still somewhat busy this weekend, so can't work on the tests rn, but can follow-up with that later, if that's ok x)

**Review by mhinz — APPROVED** [review]

**Comment by mjlbach** [comment]
@fsouza can you file a follow up PR adding a test when you have the time?

**Comment by fsouza** [comment]
> @fsouza can you file a follow up PR adding a test when you have the time?

Absolutely! Will do it on Tuesday or Wednesday.

**Comment by fsouza** [comment]
@mjlbach sent #14507.

Sorry for missing the original estimate.

**Pull request merged**